homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,263 @@
+    """Raise exception on data failure with logging."""
+    if 400 <= response.status:
+        e = ClientResponseError(
+            response.request_info,
+            response.history,
+            code=response.status,
+            headers=response.headers,
+        )
+        data = await response.text()
+        _LOGGER.error("Request failed: %s", data)
+        raise InvalidAuthError(data) from e
+
+
+registered = False
+
+WANTED_SCOPES = set(["openid", "email", "profile"])
+
+
+@AUTH_PROVIDERS.register("openid")
+class OpenIdAuthProvider(AuthProvider):
+    """Example auth provider based on hardcoded usernames and passwords."""
+
+    DEFAULT_TITLE = "OpenId Connect"
+
+    _discovery_document: Optional[Dict[str, Any]]
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Extend parent's __init__."""
+        super().__init__(*args, **kwargs)
+
+    async def async_get_discovery_document(self) -> Dict[str, Any]:
+        """Retrieve a discovery document for openid."""
+        if self._discovery_document is None:
+            session = async_get_clientsession(self.hass)
+            async with session.get(self.discovery_url) as response:
+                await raise_for_status(response)
+                self._discovery_document = cast(Dict[str, Any], await response.json())
+
+        return self._discovery_document
+
+    @property
+    def discovery_url(self) -> str:
+        """Construct discovery url based on config."""
+        return str(
+            URL(self.config[CONF_ISSUER]).with_path(".well-known/openid-configuration")
+        )
+
+    @property
+    def redirect_uri(self) -> str:
+        """Return the redirect uri."""
+        return f"{self.hass.config.api.base_url}{AUTH_CALLBACK_PATH}"  # type: ignore
+
+    async def async_login_flow(self, context: Optional[Dict]) -> LoginFlow:
+        """Return a flow to login."""
+        global registered
+        if not registered:
+            self.hass.http.register_view(OpenIdCallbackView())  # type: ignore
+        registered = True
+        return OpenIdLoginFlow(self)
+
+    async def async_retrieve_token(self, code: str) -> Dict[str, Any]:
+        """Convert a token code into an actual token."""
+        data = await self.async_get_discovery_document()
+
+        payload = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self.redirect_uri,
+            "client_id": self.config[CONF_CLIENT_ID],
+            "client_secret": self.config[CONF_CLIENT_SECRET],
+        }
+
+        session = async_get_clientsession(self.hass)
+        async with session.post(data["token_endpoint"], data=payload) as response:
+            await raise_for_status(response)
+            return cast(Dict[str, Any], await response.json())
+
+    async def async_decode_id_token(self, id_token: str) -> Dict[str, Any]:
+        """Decode a token."""
+        return jwt.decode(id_token, verify=False)
```

Comment:
Please make the user explicitly turn off verification as a configuration option so they know they are making a choice to be insecure

Reply:
Ideally the signing key is fetched from signing key assembled from the jwks_uri
https://openid.net/specs/openid-connect-discovery-1_0.html
https://tools.ietf.org/html/draft-ietf-jose-json-web-signature-41

Reply:
I must set verify=False unless i actually grab certificates. The function fails otherwise. Google mention that verification is not required if you are doing this against their servers with https. 

That might not be true for other server thou.

Reply:
Grabbing the certificates is a pain for sure.  It also makes it slower unless you cache them as well.  The upside is it makes it much harder to attack the system (https://openid.net/specs/openid-connect-core-1_0.html#Security).  It will still be _reasonably_ secure without it as long as TLS validation isn't turned off for the host.

Reply:
Fixed :)